R/crunch-data-frame.R

```diff
@@ -0,0 +1,292 @@
+    }
+    return(c(n_rows, length(ls(x))))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+names.CrunchDataFrame <- function (x) attr(x, "col_names")
+
+
+
+#' @export
+#' @rdname CrunchDataFrame
+`[.CrunchDataFrame` <- function (x, i, j, drop = TRUE) {
+    j <- grab_col_names(x, j)
+    row_inds <- grab_row_ind(x, i)
+    
+    # grab the columns
+    out <- lapply(j, function(col) {
+        col_data <- x[[col]]
+        if (is.data.frame(col_data)) {
+            return(col_data[row_inds,])
+        } else {
+            return(col_data[row_inds])
+        }
+    })
+    
+    # set names, if j is a character use that, if not grab names from the CDF
+    if (is.character(j)) {
+        names(out) <- j
+    } else {
+        names(out) <- names(x)[j]
+    }
+    
+    if (length(out) == 1 & drop) {
+        # single column, so return vector
+        out <- out[[1]]
+    } else {
+        out <- structure(out, class="data.frame", row.names=seq_along(row_inds))
+    }
+    
+    return(out)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[<-.CrunchDataFrame` <- function (x, i, j, value) {
+    j <- grab_col_names(x, j)
+    row_inds <- grab_row_ind(x, i)
+    
+    # TODO: check value length
+    
+    # split value string by the length of row indicators
+    values <- split(value, ceiling(seq_along(value)/length(row_inds)))
+    for (i in seq_along(j)) {
+        col_name <- j[i]
+        x <- set_CDF_var(col_name = col_name, row_inds = row_inds,
+                         cdf = x, value = values[[i]])
+    }
+    return(x)
+}
+
+grab_col_names <- function (x, j) {
+    if (missing(j)) {
+        # if there's no j, grab all columns
+        j <- names(x)
+    } else if (is.logical(j)) {
+        if (length(j) != ncol(x)) {
+            halt("when using a logical to subset columns, the logical vector ",
+                 "must have the same length as the number of columns.")
+        }
+        j <- names(x)[j]
+    }
+    return(j)
+}
+
+grab_row_ind <- function (x, i) {
+    if (missing(i)) {
+        # if there is no row specification, use all rows.
+        row_inds <- seq_len(nrow(x))
+    } else {
+        if (is.logical(i)) {
+            if (length(i) != nrow(x)) {
+                halt("when using a logical to subset rows, the logical vector ",
+                     "must have the same length as the number of rows")
+            }
+        } else if (!is.numeric(i)) {
+            halt("row subsetting must be done with either numeric or a logical.")
+        }
+        row_inds <- i
+    }
+    return(row_inds)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[.CrunchDataFrame` <- function (x, i) {
+    col_name <- col_name_from_index(x, i)
+    
+    return(get_CDF_var(col_name, x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[<-.CrunchDataFrame` <- function (x, i, value) {
+    col_name <- col_name_from_index(x, i)
+    
+    x <- set_CDF_var(col_name = col_name, cdf = x, value = value)
+    return(x)
```

Comment:
Not that we're golfing here, but this could easily be a single line function body, as could the one above.